### ISSUE TITLE
SMB2Client. Error on rename file: STATUS_SHARING_VIOLATION FILE_DOES_NOT_EXIST

### ISSUE THREAD

**Issue by lexamad** [TalAloni/SMBLibrary#77]
Hello!
I found the error

In connection1 open the file for reading.
In the same connection (connection1) I try to rename it, I get the error STATUS_PENDING: FILE_DOES_NOT_EXIST.
STATUS_PENDING It's okay, the file is locked by a read operation. (although FILE_DOES_NOT_EXIST is probably wrong?)
Next, I close the fileHandle of the file for reading.
I try to rename the same file again, and I get the error STATUS_SHARING_VIOLATION: FILE_DOES_NOT_EXIST.
Why? 
(Method A02)

In connection1 I open the file for reading.
In connection2 I try to rename this file, I get the error STATUS_PENDING: FILE_DOES_NOT_EXIST.
STATUS_PENDING It's okay, the file is locked by connection1.
Next, close connection1
In connection2 I try to rename the same file again, I get the error STATUS_SHARING_VIOLATION: FILE_DOES_NOT_EXIST.
Why? 
(Method B02)
 
Here is the code to reproduce
[ConsoleApp1.zip](https://github.com/TalAloni/SMBLibrary/files/6302241/ConsoleApp1.zip)

Thank you

**Comment by TalAloni** [comment]
1. The file status is only valid in some cases - your assumption that it is always meaningful is incorrect.
2. The client just pass the commands to the server and return the errors reported by the server.

Given that you are not reporting a bug with a client - but instead you have an issue with how the server handle the commands you are sending - I'm closing the issue. there is no bug here as far as I understand.

**Comment by lexamad** [comment]
Hello, I found problem
When file  is blocking another request receives 2 messages in response (in _incomingQueue 2 responses per request)

trce: LoggerDO[0]
      SmbProcessor2.Move: \Common\_ItTmp\smb_test\dd.txt to \Common\_ItTmp\smb_test\dd1.txt
dbug: LoggerDO[0]
      40362448: Request  SMBLibrary.SMB2.CreateRequest: id=6
dbug: LoggerDO[0]
      40362448: Receive  SMBLibrary.SMB2.ErrorResponse: id=6  STATUS_PENDING
dbug: LoggerDO[0]
      40362448: Receive  SMBLibrary.SMB2.ErrorResponse: id=6  STATUS_SHARING_VIOLATION

WaitForCommand only reads STATUS_PENDING,
And any subsequent request reads STATUS_SHARING_VIOLATION

you need to do something like this
if STATUS_PENDING  need to read all messages or wait until the timeout

 internal SMB2Command WaitForCommandByMessageId(ulong messageId)
        {
            const int timeOut = 5000;
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.Start();
            SMB2Command lastCommand = null;
            while (stopwatch.ElapsedMilliseconds < timeOut)
            {
                lock (_incomingQueueLock)
                {
                    int index = 0;
                    while (index < _incomingQueue.Count)
                    {
                        var cmd = _incomingQueue[index];
                        if (cmd.Header.MessageID == messageId)
                        {
                            lastCommand = cmd;
                            _incomingQueue.RemoveAt(index);
                            if (cmd is ErrorResponse && cmd.Header.Status == NTStatus.STATUS_PENDING)
                                continue;
                        }
                        index++;
                    }
                }
                _incomingQueueEventHandle.WaitOne(100);
            }
            return lastCommand;
        }

**Comment by TalAloni** [comment]
It seems like a violation of the SMB2 protocol when a single synchronous request is answered by two responses. I was in disbelief but I was able to run your program and saw it too (Server 2008 R2).
With that said, I'm not sure I can qualify what you are doing as a valid behavior: you have a client opening two handles to the same file and then attempting to rename the file - knowing full well that the operation cannot succeed. so I'm not sure this specific example is valid and worthy of investigation. (unrelated: note that you are not logging off before disconnection)

**Comment by TalAloni** [comment]
Thanks,
After reviewing the specifications again, I have made the necessary correction to handling of async responses.

**Comment by lexamad** [comment]
Ok, thank you
One question - why are you check by commandName, not a MessageID?
for example
SMB2Command response = m_client.WaitForCommand(SMB2CommandName.Create);
or
SMB2Command response = m_client.WaitForCommand(request.Header.MessageID);

Probably, with an asynchronous request by MessageID, you can get the answer more clearly?
for example, as I think, when we opened a large file for reading and read it in parts, in parallel we opened another file - and you can catch messages that are not intended for the second file

**Comment by TalAloni** [comment]
lexamad, your observation is correct, there is no good reason not to use MessageID. (SMB2Client was based on the SMB1Client, and SMB1 does not have MessageID - so I guess that's the reason).

**Comment by lexamad** [comment]
Will you correct the messageID? When to wait for a package in nuget? :)

**Comment by TalAloni** [comment]
Both corrections are now in v1.4.6.1 published on NuGet.org

**Comment by lexamad** [comment]
Thank you!

**Comment by lexamad** [comment]
SessionID must also be used with MessageID, messages may be from different sessions?

**Comment by TalAloni** [comment]
Are you asking or telling?

**Comment by lexamad** [comment]
You can do Connect, OpenFile, Disconnect, and receive command may remain in _incomingQueue, we do Connect again - and it may so happen that the messageId will be the same for different sessions. Or need clearing _incomingQueue on Connect

And yet, maybe you need to save the lastCommand so that you can transfer it later if there was a timeout?

   internal SMB2Command WaitForCommand(ulong messageID)
        {
            const int TimeOut = 5000;
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.Start();
            SMB2Command lastCommand = null;
            while (stopwatch.ElapsedMilliseconds < TimeOut)
            {
                lock (_incomingQueueLock)
                {
                    for (int index = 0; index < _incomingQueue.Count; index++)
                    {
                        SMB2Command command = _incomingQueue[index];

                        if (command.Header.MessageID == messageID)
                        {
                            _incomingQueue.RemoveAt(index);
                            if (command.Header.IsAsync && command.Header.Status == NTStatus.STATUS_PENDING)
                            {
                                index--;
                                lastCommand = command;
                                continue;
                            }
                            return command;
                        }
                    }
                }
                _incomingQueueEventHandle.WaitOne(100);
            }
            return lastCommand;
        }

**Comment by TalAloni** [comment]
I find the scenario you are suggesting a bit of a stretch as it also require the server to behave in a certain way.
Also, this is outside the scope of the original issue - if you have another bug to report - please create another issue.

**Comment by lexamad** [comment]
:) Thank you
